moonraker/components/sensor.py

```diff
@@ -0,0 +1,232 @@
+        )
+
+        for section in prefix_sections:
+            cfg = config[section]
+
+            try:
+                try:
+                    _, name = cfg.get_name().split(maxsplit=1)
+                except ValueError:
+                    raise cfg.error(f"Invalid section name: {cfg.get_name()}")
+
+                logging.info(f"Configuring sensor: {name}")
+
+                sensor_type: str = cfg.get("type", "mqtt")
+                sensor_class: Optional[
+                    Type[BaseSensor]
+                ] = self.__sensor_types.get(sensor_type.upper(), None)
+                if sensor_class is None:
+                    raise config.error(
+                        f"Unsupported sensor type: {sensor_type}"
+                    )
+
+                self.sensors[name] = sensor_class(name=name, cfg=cfg)
+            except Exception as e:
+                # Ensures that configuration errors are shown to the user
+                self.server.add_warning(
+                    f"Failed to configure sensor [{cfg.get_name()}]\n{e}"
+                )
+                continue
+
+    async def component_init(self) -> None:
+        try:
+            logging.debug("Initializing sensor component")
+            event_loop = self.server.get_event_loop()
+            cur_time = event_loop.get_loop_time()
+            endtime = cur_time + 120.0
+
+            uninitialized_sensors = list(self.sensors.values())
+            while cur_time < endtime:
+                failed_sensors: List[BaseSensor] = []
+                for sensor in uninitialized_sensors:
+                    ret = sensor.initialize()
+                    if ret is not None:
+                        await ret
+                    if sensor.error_state is not None:
+                        failed_sensors.append(sensor)
+                if not failed_sensors:
+                    logging.debug("All sensors have been initialized")
+                    return
+
+                uninitialized_sensors = failed_sensors
+                await asyncio.sleep(2.0)
+                cur_time = event_loop.get_loop_time()
+
+            for sensor in failed_sensors:
+                msg = (
+                    f"Sensor {sensor.config.name} is not available: "
+                    f"{sensor.error_state}"
+                )
+                logging.warning(msg)
+                self.server.add_warning(msg)
+
+        except Exception as e:
+            logging.exception(e)
+
+    async def _handle_sensor_data_request(
+        self, web_request: WebRequest
+    ) -> Dict[str, Dict[str, Any]]:
+        data = {}
+        for sensor in self.sensors.values():
+            data[sensor.config.id] = {
+                "source": sensor.config.source,
+                "name": sensor.config.name,
+                "unit_of_measurement": sensor.config.unit,
+                "accuracy_decimals": sensor.config.accuracy_decimals,
+                "measurements": list(sensor.values),
+            }
```

Comment:
With the above recommendations this would probably report a result something like the following:
```python
{
    "bme280": {
        "friendly_name": "bme_280",
        "values": {
            "temperature": 21.4,
            "humidity": "33"
        }
    }
}
```

Likewise, an endpoint that reports the store might look like the following:
```python
{
    "bme280": {
        "temperature": [21.4, 22.0, 21.8]
        "humidity": ["33.1, 33.1, 33.0]
    }
}
```